proposals/0016-simd-application-fees.md

```diff
@@ -0,0 +1,278 @@
+
+*Passing the application fees in the instruction for each account and validating in the dapp*. A `PayApplicationFee`
+instruction is like irrevocable transfer instruction and will do the transfer even if the transaction fails. A new
+instruction `CheckApplicationFee` will check if the application fee for an account has been paid.
+
+    1. High degree of flexibility for programs (**++**)
+    2. No need to save the application fees on the ledger. (**++**)
+    3. Each transaction has to include additional instructions so it will break existing dapp interfaces (**-**)
+    4. Account structure does not need to be modified (**++**)
+    5. Does not prevent any denial of service attack on a smart contract (**-**)
+    6. Each dapp has to implement a cpi to check if a transaction has paid app fees
+
+If existing dapps like openbook want to implement application fees then all the dapps depending on openbook 
+also have to do additional development. The checks on the application fees will be responsibility of dapp developers.
+The plus point is that we do not have to touch the core solana structure to store the application fees.
+It is also more easier to calculate total fees.
+
+### A new application fee program
+
+We add a new native solana program called application fees program with program id.
+
+```
+App1icationFees1111111111111111111111111111
+```
+
+This native program will be used to check application fee for an account, to intialize rebates
+initaited by the account authority, and a special instruction by which fee payer accepts
+amount of lamports they are willing to pay as application fees per account.
+
+#### PayApplicationFees Instruction
+
+With this instruction, the fee payer accepts to pay application fees specifying the amount.
+This instruction **MUST** be included in the transaction that interacts with dapps having application fees.
+This instruction is like an irrevokable transfer if the payer has enough funds
+i.e even if the transaction fails the payer will end up paying.
+If this instruction is added then even if the dapp does not check the application fees the payer ends
+up paying.
+If the payer does not have enough balance the transaction fails with error `InsufficientFunds`.
+
+It requires:
+
+Accounts :
+* List of accounts that need an application fees.
+
+Argument : Corresponding list of application fees for each account in `Accounts`.
+
+The arguments list must be of same length as number of accounts.
+The index of fees and account should match.
+The account fees for each account is 8 bytes represented in rust type `u64`
+
+#### CheckApplicationFees Instruction
+
+This instruction will check if an application fee for an account is paid.
+It requires :
+
+* Account where fees are paid.
+
+Argument: required fees in lamport (u64).
+
+If application fees are not paid or are paid insufficiently this instruction will return
+an error. The idea is dapp developer uses this instruction to check if the required fees
```

Comment:
In practice, how do you expect these fees to be set? Actively and dynamically or set-and-forget? Should they respond to market conditions? If so, does this require a privileged user to operate the program in perpetuity?

Reply:
To set the fees we have to redeploy the smart contract, because the fees are checked on the dapp side using `CheckApplicationFees` instruction. So i would say it rather set-and-forget.

Reply:
Do you expect the fees to change over time or to remain static? In practice how should developers think about how to set application fees?

Reply:
I would say developer should look it as a `static` fees not dependent on cluster congestion etc.
Though they can update it if they want to find reasonable fees that suite their requirements.